src/pymmcore_plus/mda/_engine.py

```diff
@@ -54,11 +69,46 @@ def setup_event(self, event: MDAEvent) -> None:
         self._mmc.waitForSystem()
 
     def exec_event(self, event: MDAEvent) -> Any:
-        """Execute an individual event and return the image data."""
-        # TODO: add non-aquisition event-specific logic here later
+        """Execute an individual event."""
+        action = getattr(event, "action", None)
+
+        # execute hardware autofocus
+        if isinstance(action, HardwareAutofocus) and event.z_pos is not None:
+            # get position index
+            p_idx = event.index.get("p", None)
+            # run autofocus and get the correction to apply to each z position
+            self._z_correction[p_idx] = self._execute_autofocus(action)
+            return None
+
+        # acquire an image and emit the image data
         self._mmc.snapImage()
         return EventPayload(image=self._mmc.getImage())
 
+    def _execute_autofocus(self, action: HardwareAutofocus) -> float:
+        """Perform the hardware autofocus.
+
+        Returns the z correction to apply to each z position.
+        """
+        self._mmc.setPosition(
+            action.autofocus_device_name,
+            action.autofocus_motor_offset,
+        )
+        self._mmc.waitForSystem()
+
+        current_z = self._mmc.getZPosition()
+
+        @retry(exceptions=RuntimeError, tries=action.max_retries, logger=logger.warning)
+        def _full_focus() -> float:
+            self._mmc.fullFocus()
+            self._mmc.waitForSystem()
+            return self._mmc.getZPosition() - current_z
+
+        try:
+            return _full_focus()
+        except RuntimeError:
+            logger.warning("Warning: Hardware autofocus failed.")
+            return 0.0
```

Comment:
you could also put this try/except above in `exec_event`

```python
try:
    self._z_correction[p_idx] = self._execute_autofocus(action)
except RuntimeWarning:
    logger.warning("Warning: Hardware autofocus failed.")
```

which would prevent you from overriding a previous correction (that might indeed be closer to the truth) with the `0.0` value